Pods pending due to insufficient OIR should get scheduled once sufficient OIR becomes available.

ConnorDoyle opened the pull request:
This appears to be a regression since v1.5.0 in scheduler behavior for opaque integer resources, reported in https://github.com/kubernetes/kubernetes/issues/41861.

- [X] Add failing e2e test to trigger the regression
- [x] Restore previous behavior (pods pending due to insufficient OIR get scheduled once sufficient OIR becomes available.)

ConnorDoyle committed:
Added missing OIR e2e test: sched after release.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41870)
<!-- Reviewable:end -->

timothysc commented:
@ConnorDoyle let us know when it's ready for review.

ConnorDoyle commented:
@timothysc will do -- quick question regarding the e2e tests though. They may not have prevented the regression since we were missing an important case. However, in the time since they were written thing changed that causes the existing cases to fail. What's the policy for which tests run as part of CI, and this time around could we leave the `[Feature:` suffix off?

timothysc commented:
So "[Feature:" tests are omitted as part of the default CI, we should enable for coverage.

aveshagarwal commented:
@kubernetes/sig-scheduling-pr-reviews

ConnorDoyle committed:
Fix unbounded growth of cached OIRs in sched cache

- Added schedulercache.Resource.SetOpaque helper.
- Amend kubelet allocatable sync so that when OIRs are removed from capacity
  they are also removed from allocatable.
- Fixes #41861.

ConnorDoyle commented:
@timothysc verified this fix with local cluster. Still messing with the e2e's but this is ready for 1st pass review when you get a minute.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -38,7 +38,7 @@ import (
 )
 
-var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", func() {
+var _ = framework.KubeDescribe("Opaque resources", func() {
```
This likely needs to marked as [Serial] b/c you're mucking the nodes.

ConnorDoyle replied:
Added [Serial] tag.

ConnorDoyle replied:
...and removed [Serial] tag.

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -215,9 +217,65 @@ var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", fun
-				e.Reason == "FailedScheduling" &&
-				strings.Contains(e.Message, "failed to fit in any node")
+			return strings.HasPrefix(e.Name, pod.Name) &&
+				e.Reason == "OutOfpod.alpha.kubernetes.io/opaque-int-resource-foo"
```
Is this really the reason the user will see when they list events?

ConnorDoyle replied:
Not after the fix (test needs update). This is what happens when a pod gets bound to a node with less than the necessary resource available. This is what the person who reported the bug saw. The Kubelet sets the pod phase to Failed and this is what shows up in kubectl.

Actually that's a separate question, and a separate issue -- what should happen in that case? Do users expect the pod to bounce off of that node and get scheduled or do they expect this terminal state?

timothysc commented on test/e2e/opaque_resource.go:
```diff
@@ -215,9 +217,65 @@ var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", fun
+		Expect(success).To(Equal(true))
+	})
+
+	It("should schedule pods that initially do not fit after enough  opaque integer resources are freed.", func() {
```
two spaces after enough.

ConnorDoyle replied:
Thanks, fixed.

timothysc commented:
Maybe I'm missing it, but are you cleaning up the muck(add of oir) on the nodes and ensuring all pods are cleaned at the end of your tests?

ConnorDoyle commented:
@timothysc I added an AfterEach block to remove the OIRs following the test. IIUC the whole namespace gets destroyed after the test so I didn't worry about orphaning pods. If there's an example somewhere in the e2e suite you'd like me to follow just send a pointer.

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -505,10 +505,8 @@ func GetResourceRequest(pod *v1.Pod) *schedulercache.Resource {
 			default:
```
feels like this method needs a header comment `// GetResourceRequest calculates maximal individual init containers (since they're serial) and sum of standard pod containers (since theyre running together), and takes the max of the entire set.`

ConnorDoyle replied:
Agree, will add this _en passent_.

ConnorDoyle replied:
Added a banner comment.

aveshagarwal commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -90,6 +90,14 @@ func (r *Resource) AddOpaque(name v1.ResourceName, quantity int64) {
+		r.OpaqueIntResources = map[v1.ResourceName]int64{}
+	}
+	r.OpaqueIntResources[name] = quantity
+}
```
Wonder if AddOpaque and SetOpaque could be combined in one function:

```
if r.OpaqueIntResources == nil {
 		r.OpaqueIntResources = map[v1.ResourceName]int64{}
                r.OpaqueIntResources[name] = quantity
 	} else {
  	        r.OpaqueIntResources[name] += quantity
}
```

ConnorDoyle replied:
It would be nice but I don't think so without a larger refactor. The call site within SetNode requires an idempotent operation.

jayunit100 commented on plugin/pkg/scheduler/schedulercache/node_info.go:
```diff
@@ -90,6 +90,14 @@ func (r *Resource) AddOpaque(name v1.ResourceName, quantity int64) {
 	r.OpaqueIntResources[name] += quantity
 }
 
+func (r *Resource) SetOpaque(name v1.ResourceName, quantity int64) {
```
nit: to avoid duplication of lazy init i think `AddOpaque` should only do the calculation for adding and then call `SetOpaque`?

ConnorDoyle replied:
OK, we can do that.

jayunit100 replied:
Thanks looks good

jayunit100 commented on plugin/pkg/scheduler/algorithm/predicates/predicates.go:
```diff
@@ -505,10 +505,8 @@ func GetResourceRequest(pod *v1.Pod) *schedulercache.Resource {
-					result.AddOpaque(rName, int64(0))
 					if value > result.OpaqueIntResources[rName] {
-						result.OpaqueIntResources[rName] = value
+						result.SetOpaque(rName, value)
```
should we update unit test for the fix to ?

ConnorDoyle replied:
Which unit tests do you have in mind? We could add something for SetOpaque.

jayunit100 replied:
Yup;

aveshagarwal commented on test/e2e/opaque_resource.go:
```diff
@@ -63,6 +63,10 @@ var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", fun
 	})
 
+	AfterEach(func() {
+		removeOpaqueResource(f, node.Name, opaqueResName)
```
As you are calling it here, you could remove the same in BeforeEach.

ConnorDoyle replied:
Good catch. I'll try removing it after things are passing again.

aveshagarwal commented on test/e2e/opaque_resource.go:
```diff
@@ -373,3 +401,20 @@ func observeEventAfterAction(f *framework.Framework, eventPredicate func(*v1.Eve
+	}
+}
+
+func scheduleFailure(podName string) func(*v1.Event) bool {
```
why `!scheduleSuccess` is not enough to check for failure?

ConnorDoyle replied:
Because then we would have to wait to an event _not_ to appear which is much slower.

aveshagarwal commented on test/e2e/opaque_resource.go:
```diff
@@ -38,7 +38,7 @@ import (
 )
 
-var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", func() {
+var _ = framework.KubeDescribe("Opaque resources [Serial]", func() {
```
why should it be serial? If these tests are only adding/removing/modifying OIRs to nodes, why would it impact other tests that dont deal with OIRs? IOW why adding additional resource types to nodes would impact other tests that dont deal with the new types?

ConnorDoyle replied:
It was not serial, then @timothysc suggested making it serial. I'm afraid I don't understand the implications enough to make an educated decision. Does it control parallelism of just this group of tests or does it span the entire e2e framework?

timothysc replied:
The Serial was to allow for node cleanup without potential conflicts.  It's the safest thing you can do given that it was failing e2es.

ConnorDoyle replied:
Alright. I think it was/is failing for other reasons. Once it passes we an try removing the serial tag.

aveshagarwal replied:
Without serial, this group would run in parallel to other groups. And I dont see this group affecting other groups when run in parallel.

ConnorDoyle replied:
gce e2e passed; removed [Serial] now.

ConnorDoyle commented:
@k8s-bot unit test this
@k8s-bot verify test this
@k8s-bot cvm gce e2e test this

timothysc commented:
poke, once it's all green and ready.. this is a bug so your in the clear for merging.

ConnorDoyle commented:
Still struggling with the e2e... when ginkgo times out there isn't enough output to help with debugging -- not even sure it's this test that is causing the test suite to hang long enough to timeout.

```
I0301 12:56:25.896] 	 -------------------------------------------------------------------
I0301 12:56:25.896] 	|                                                                   |
I0301 12:56:25.896] 	|  Ginkgo timed out waiting for all parallel nodes to report back!  |
I0301 12:56:25.896] 	|                                                                   |
I0301 12:56:25.896] 	 -------------------------------------------------------------------
I0301 12:56:25.897] 
```

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

aveshagarwal commented:
@ConnorDoyle I see these in your verification logs:

```
W0301 12:26:54.141] test/e2e/opaque_resource.go:264: arg foundCap for printf verb %s of wrong type: bool
W0301 12:26:54.141] test/e2e/opaque_resource.go:284: arg foundCap for printf verb %s of wrong type: bool
```

ConnorDoyle commented:
Thanks @aveshagarwal!

ConnorDoyle commented:
@k8s-bot unit test this

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kops aws e2e test this
@k8s-bot non-cri e2e test this

ConnorDoyle commented:
@k8s-bot cvm gce e2e test this

ConnorDoyle commented:
@k8s-bot gce etcd3 e2e test this

ConnorDoyle commented:
@k8s-bot gce etcd3 e2e test this

ConnorDoyle commented:
@timothysc this is (finally) green again, sorry for the delay. There were multiple fixes required and it took a while to figure out what was happening on the Kubelet side.

ConnorDoyle commented:
@k8s-bot gce etcd3 e2e test this

ConnorDoyle commented:
@k8s-bot gce etcd3 e2e test this

timothysc commented:
@k8s-bot gci gce e2e test this

timothysc reviewed (commented):
So on initial inspection looks pretty clean, but your comment re: AfterEach raises a concern.  Do you know the reason it fails?

ConnorDoyle commented on test/e2e/opaque_resource.go:
```diff
@@ -59,11 +59,19 @@ var _ = framework.KubeDescribe("Opaque resources [Feature:OpaqueResources]", fun
 		addOpaqueResource(f, node.Name, opaqueResName)
 	})
 
+	// TODO: The suite times out if removeOpaqueResource is called as part of
```
No, I am stumped on why this causes a problem. Open to any ideas.

timothysc replied:
/cc @kubernetes/test-infra-reviewers any insights here?

timothysc replied:
So I looked at anti-affinity and it does both, so it looks like they just patched and didn't bother to clearly define a happy path.

timothysc commented:
I don't want to hold this up, lgtm.  

We need to revisit several scheduling e2es to determine the proper way to cleanup.  
@ConnorDoyle could you open an issue re: your comment above please.

ConnorDoyle commented:
@timothysc do we need to do anything special to get this cherry-picked into 1.6?

timothysc commented:
Not yet, I believe the branches are still fast-forward.

ConnorDoyle commented:
@timothysc created issue #42590 to track the AfterEach weirdness.

dchen1107 commented:
https://github.com/kubernetes/kubernetes/issues/42597

timothysc commented:
welp already reverted.

ConnorDoyle commented:
Better to burn out than to fade away. Preparing a new PR with the test disabled; will re-enable them in follow-ups.
